File: qt/gen_qglyphrun.go
package qt

import (
	"unsafe"
)

type QGlyphRun__GlyphRunFlag int

const (
	QGlyphRun__Overline      QGlyphRun__GlyphRunFlag = 1
	QGlyphRun__Underline     QGlyphRun__GlyphRunFlag = 2
	QGlyphRun__StrikeOut     QGlyphRun__GlyphRunFlag = 4
	QGlyphRun__RightToLeft   QGlyphRun__GlyphRunFlag = 8
	QGlyphRun__SplitLigature QGlyphRun__GlyphRunFlag = 16
)

type QGlyphRun struct {
	h          uintptr
	isSubclass bool
}

// NewQGlyphRun constructs a new QGlyphRun object.
func NewQGlyphRun() *QGlyphRun {

	ret := newQGlyphRun(QGlyphRun_new())
	ret.isSubclass = true
	return ret
}

// NewQGlyphRun2 constructs a new QGlyphRun object.
func NewQGlyphRun2(other *QGlyphRun) *QGlyphRun {

	ret := newQGlyphRun(QGlyphRun_new2(other.cPointer()))
	ret.isSubclass = true
	return ret
}

func (this *QGlyphRun) OperatorAssign(other *QGlyphRun) {
	QGlyphRun_OperatorAssign(this.h, other.cPointer())
}

func (this *QGlyphRun) Swap(other *QGlyphRun) {
	QGlyphRun_Swap(this.h, other.cPointer())
}

func (this *QGlyphRun) RawFont() *QRawFont {
	_goptr := newQRawFont(QGlyphRun_RawFont(this.h))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QGlyphRun) SetRawFont(rawFont *QRawFont) {
	QGlyphRun_SetRawFont(this.h, rawFont.cPointer())
}

func (this *QGlyphRun) SetRawData(glyphIndexArray *uint, glyphPositionArray *QPointF, size int) {
	QGlyphRun_SetRawData(this.h, (*uint)(unsafe.Pointer(glyphIndexArray)), glyphPositionArray.cPointer(), (int)(size))
}

func (this *QGlyphRun) GlyphIndexes() []uint {
	var _ma struct_miqt_array = QGlyphRun_GlyphIndexes(this.h)
	_ret := make([]uint, int(_ma.len))
	_outCast := (*[0xffff]uint)(unsafe.Pointer(_ma.data)) // hey ya
	for i := 0; i < int(_ma.len); i++ {
		_ret[i] = (uint)(_outCast[i])
	}
	return _ret
}

func (this *QGlyphRun) SetGlyphIndexes(glyphIndexes []uint) {
	glyphIndexes_CArray := (*[0xffff]uint)(malloc(size_t(8 * len(glyphIndexes))))
	defer free(unsafe.Pointer(glyphIndexes_CArray))
	for i := range glyphIndexes {
		glyphIndexes_CArray[i] = (uint)(glyphIndexes[i])
	}
	glyphIndexes_ma := struct_miqt_array{len: size_t(len(glyphIndexes)), data: unsafe.Pointer(glyphIndexes_CArray)}
	QGlyphRun_SetGlyphIndexes(this.h, glyphIndexes_ma)
}

func (this *QGlyphRun) Positions() []QPointF {
	var _ma struct_miqt_array = QGlyphRun_Positions(this.h)
	_ret := make([]QPointF, int(_ma.len))
	_outCast := (*[0xffff]*QPointF)(unsafe.Pointer(_ma.data)) // hey ya
	for i := 0; i < int(_ma.len); i++ {
		_lv_goptr := newQPointF(_outCast[i])
		_lv_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
		_ret[i] = *_lv_goptr
	}
	return _ret
}

func (this *QGlyphRun) SetPositions(positions []QPointF) {
	positions_CArray := (*[0xffff]*QPointF)(malloc(size_t(8 * len(positions))))
	defer free(unsafe.Pointer(positions_CArray))
	for i := range positions {
		positions_CArray[i] = positions[i].cPointer()
	}
	positions_ma := struct_miqt_array{len: size_t(len(positions)), data: unsafe.Pointer(positions_CArray)}
	QGlyphRun_SetPositions(this.h, positions_ma)
}

func (this *QGlyphRun) Clear() {
	QGlyphRun_Clear(this.h)
}

func (this *QGlyphRun) OperatorEqual(other *QGlyphRun) bool {
	return (bool)(QGlyphRun_OperatorEqual(this.h, other.cPointer()))
}

func (this *QGlyphRun) OperatorNotEqual(other *QGlyphRun) bool {
	return (bool)(QGlyphRun_OperatorNotEqual(this.h, other.cPointer()))
}

func (this *QGlyphRun) SetOverline(overline bool) {
	QGlyphRun_SetOverline(this.h, (bool)(overline))
}

func (this *QGlyphRun) Overline() bool {
	return (bool)(QGlyphRun_Overline(this.h))
}

func (this *QGlyphRun) SetUnderline(underline bool) {
	QGlyphRun_SetUnderline(this.h, (bool)(underline))
}

func (this *QGlyphRun) Underline() bool {
	return (bool)(QGlyphRun_Underline(this.h))
}

func (this *QGlyphRun) SetStrikeOut(strikeOut bool) {
	QGlyphRun_SetStrikeOut(this.h, (bool)(strikeOut))
}

func (this *QGlyphRun) StrikeOut() bool {
	return (bool)(QGlyphRun_StrikeOut(this.h))
}

func (this *QGlyphRun) SetRightToLeft(on bool) {
	QGlyphRun_SetRightToLeft(this.h, (bool)(on))
}

func (this *QGlyphRun) IsRightToLeft() bool {
	return (bool)(QGlyphRun_IsRightToLeft(this.h))
}

func (this *QGlyphRun) SetFlag(flag GlyphRunFlag) {
	QGlyphRun_SetFlag(this.h, flag)
}

func (this *QGlyphRun) SetFlags(flags GlyphRunFlags) {
	QGlyphRun_SetFlags(this.h, flags)
}

func (this *QGlyphRun) Flags() GlyphRunFlags {
	xxxxxxxxx
}

func (this *QGlyphRun) SetBoundingRect(boundingRect *QRectF) {
	QGlyphRun_SetBoundingRect(this.h, boundingRect.cPointer())
}

func (this *QGlyphRun) BoundingRect() *QRectF {
	_goptr := newQRectF(QGlyphRun_BoundingRect(this.h))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QGlyphRun) StringIndexes() []int64 {
	var _ma struct_miqt_array = QGlyphRun_StringIndexes(this.h)
	_ret := make([]int64, int(_ma.len))
	_outCast := (*[0xffff]ptrdiff_t)(unsafe.Pointer(_ma.data)) // hey ya
	for i := 0; i < int(_ma.len); i++ {
		_ret[i] = (int64)(_outCast[i])
	}
	return _ret
}

func (this *QGlyphRun) SetStringIndexes(stringIndexes []int64) {
	stringIndexes_CArray := (*[0xffff]ptrdiff_t)(malloc(size_t(8 * len(stringIndexes))))
	defer free(unsafe.Pointer(stringIndexes_CArray))
	for i := range stringIndexes {
		stringIndexes_CArray[i] = (ptrdiff_t)(stringIndexes[i])
	}
	stringIndexes_ma := struct_miqt_array{len: size_t(len(stringIndexes)), data: unsafe.Pointer(stringIndexes_CArray)}
	QGlyphRun_SetStringIndexes(this.h, stringIndexes_ma)
}

func (this *QGlyphRun) SetSourceString(sourceString string) {
	sourceString_ms := struct_miqt_string{}
	sourceString_ms.data = CString(sourceString)
	sourceString_ms.len = size_t(len(sourceString))
	defer free(unsafe.Pointer(sourceString_ms.data))
	QGlyphRun_SetSourceString(this.h, sourceString_ms)
}

func (this *QGlyphRun) SourceString() string {
	var _ms struct_miqt_string = QGlyphRun_SourceString(this.h)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QGlyphRun) IsEmpty() bool {
	return (bool)(QGlyphRun_IsEmpty(this.h))
}

func (this *QGlyphRun) SetFlag2(flag GlyphRunFlag, enabled bool) {
	QGlyphRun_SetFlag2(this.h, flag, (bool)(enabled))
}

File: qt/gen_qglyphrun.cpp
// +build ignore

#include <QGlyphRun>
#include <QList>
#include <QPointF>
#include <QRawFont>
#include <QRectF>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qglyphrun.h>
#include "gen_qglyphrun.h"

#ifndef _Bool
#define _Bool bool
#endif

void _GUID_Delete(_GUID* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<_GUID*>( self );
	} else {
		delete self;
	}
}

void type_info_Delete(type_info* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<type_info*>( self );
	} else {
		delete self;
	}
}

QGlyphRun* QGlyphRun_new() {
	return new QGlyphRun();
}

QGlyphRun* QGlyphRun_new2(QGlyphRun* other) {
	return new QGlyphRun(*other);
}

void QGlyphRun_OperatorAssign(QGlyphRun* self, QGlyphRun* other) {
	self->operator=(*other);
}

void QGlyphRun_Swap(QGlyphRun* self, QGlyphRun* other) {
	self->swap(*other);
}

QRawFont* QGlyphRun_RawFont(const QGlyphRun* self) {
	return new QRawFont(self->rawFont());
}

void QGlyphRun_SetRawFont(QGlyphRun* self, QRawFont* rawFont) {
	self->setRawFont(*rawFont);
}

void QGlyphRun_SetRawData(QGlyphRun* self, const unsigned int* glyphIndexArray, QPointF* glyphPositionArray, int size) {
	self->setRawData(static_cast<const quint32*>(glyphIndexArray), glyphPositionArray, static_cast<int>(size));
}

struct miqt_array /* of unsigned int */  QGlyphRun_GlyphIndexes(const QGlyphRun* self) {
	QList<quint32> _ret = self->glyphIndexes();
	// Convert QList<> from C++ memory to manually-managed C memory
	unsigned int* _arr = static_cast<unsigned int*>(malloc(sizeof(unsigned int) * _ret.length()));
	for (size_t i = 0, e = _ret.length(); i < e; ++i) {
		_arr[i] = _ret[i];
	}
	struct miqt_array _out;
	_out.len = _ret.length();
	_out.data = static_cast<void*>(_arr);
	return _out;
}

void QGlyphRun_SetGlyphIndexes(QGlyphRun* self, struct miqt_array /* of unsigned int */  glyphIndexes) {
	QList<quint32> glyphIndexes_QList;
	glyphIndexes_QList.reserve(glyphIndexes.len);
	unsigned int* glyphIndexes_arr = static_cast<unsigned int*>(glyphIndexes.data);
	for(size_t i = 0; i < glyphIndexes.len; ++i) {
		glyphIndexes_QList.push_back(static_cast<unsigned int>(glyphIndexes_arr[i]));
	}
	self->setGlyphIndexes(glyphIndexes_QList);
}

struct miqt_array /* of QPointF* */  QGlyphRun_Positions(const QGlyphRun* self) {
	QList<QPointF> _ret = self->positions();
	// Convert QList<> from C++ memory to manually-managed C memory
	QPointF** _arr = static_cast<QPointF**>(malloc(sizeof(QPointF*) * _ret.length()));
	for (size_t i = 0, e = _ret.length(); i < e; ++i) {
		_arr[i] = new QPointF(_ret[i]);
	}
	struct miqt_array _out;
	_out.len = _ret.length();
	_out.data = static_cast<void*>(_arr);
	return _out;
}

void QGlyphRun_SetPositions(QGlyphRun* self, struct miqt_array /* of QPointF* */  positions) {
	QList<QPointF> positions_QList;
	positions_QList.reserve(positions.len);
	QPointF** positions_arr = static_cast<QPointF**>(positions.data);
	for(size_t i = 0; i < positions.len; ++i) {
		positions_QList.push_back(*(positions_arr[i]));
	}
	self->setPositions(positions_QList);
}

void QGlyphRun_Clear(QGlyphRun* self) {
	self->clear();
}

bool QGlyphRun_OperatorEqual(const QGlyphRun* self, QGlyphRun* other) {
	return (*self == *other);
}

bool QGlyphRun_OperatorNotEqual(const QGlyphRun* self, QGlyphRun* other) {
	return (*self != *other);
}

void QGlyphRun_SetOverline(QGlyphRun* self, bool overline) {
	self->setOverline(overline);
}

bool QGlyphRun_Overline(const QGlyphRun* self) {
	return self->overline();
}

void QGlyphRun_SetUnderline(QGlyphRun* self, bool underline) {
	self->setUnderline(underline);
}

bool QGlyphRun_Underline(const QGlyphRun* self) {
	return self->underline();
}

void QGlyphRun_SetStrikeOut(QGlyphRun* self, bool strikeOut) {
	self->setStrikeOut(strikeOut);
}

bool QGlyphRun_StrikeOut(const QGlyphRun* self) {
	return self->strikeOut();
}

void QGlyphRun_SetRightToLeft(QGlyphRun* self, bool on) {
	self->setRightToLeft(on);
}

bool QGlyphRun_IsRightToLeft(const QGlyphRun* self) {
	return self->isRightToLeft();
}

void QGlyphRun_SetFlag(QGlyphRun* self, GlyphRunFlag flag) {
	self->setFlag(flag);
}

void QGlyphRun_SetFlags(QGlyphRun* self, GlyphRunFlags flags) {
	self->setFlags(flags);
}

GlyphRunFlags QGlyphRun_Flags(const QGlyphRun* self) {
	return self->flags();
}

void QGlyphRun_SetBoundingRect(QGlyphRun* self, QRectF* boundingRect) {
	self->setBoundingRect(*boundingRect);
}

QRectF* QGlyphRun_BoundingRect(const QGlyphRun* self) {
	return new QRectF(self->boundingRect());
}

struct miqt_array /* of ptrdiff_t */  QGlyphRun_StringIndexes(const QGlyphRun* self) {
	QList<qsizetype> _ret = self->stringIndexes();
	// Convert QList<> from C++ memory to manually-managed C memory
	ptrdiff_t* _arr = static_cast<ptrdiff_t*>(malloc(sizeof(ptrdiff_t) * _ret.length()));
	for (size_t i = 0, e = _ret.length(); i < e; ++i) {
		_arr[i] = _ret[i];
	}
	struct miqt_array _out;
	_out.len = _ret.length();
	_out.data = static_cast<void*>(_arr);
	return _out;
}

void QGlyphRun_SetStringIndexes(QGlyphRun* self, struct miqt_array /* of ptrdiff_t */  stringIndexes) {
	QList<qsizetype> stringIndexes_QList;
	stringIndexes_QList.reserve(stringIndexes.len);
	ptrdiff_t* stringIndexes_arr = static_cast<ptrdiff_t*>(stringIndexes.data);
	for(size_t i = 0; i < stringIndexes.len; ++i) {
		stringIndexes_QList.push_back(static_cast<QIntegerForSizeof<std::size_t>::Signed>(stringIndexes_arr[i]));
	}
	self->setStringIndexes(stringIndexes_QList);
}

void QGlyphRun_SetSourceString(QGlyphRun* self, struct miqt_string sourceString) {
	QString sourceString_QString = QString::fromUtf8(sourceString.data, sourceString.len);
	self->setSourceString(sourceString_QString);
}

struct miqt_string QGlyphRun_SourceString(const QGlyphRun* self) {
	QString _ret = self->sourceString();
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

bool QGlyphRun_IsEmpty(const QGlyphRun* self) {
	return self->isEmpty();
}

void QGlyphRun_SetFlag2(QGlyphRun* self, GlyphRunFlag flag, bool enabled) {
	self->setFlag(flag, enabled);
}

void QGlyphRun_Delete(QGlyphRun* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<QGlyphRun*>( self );
	} else {
		delete self;
	}
}

